### ISSUE TITLE
LobbyTransaction can't be configured

### ISSUE THREAD

**Issue by boringcactus** [ldesgoui/discord_game_sdk#36]
I'm trying to replace my hand-rolled TCP stack with Discord lobbies, but I'm having issues using `LobbyTransaction`. In short, the code that doesn't work is

```rust
let mut txn = LobbyTransaction::new();
txn.locked(true);
discord.update_lobby(id, txn, callback);
```

This is rejected with
```
error[E0505]: cannot move out of `txn` because it is borrowed
  --> src/main.rs:28:17
   |
27 |     txn.locked(true);
   |     --- borrow of `txn` occurs here
28 |     consume_txn(txn);
   |                 ^^^
   |                 |
   |                 move out of `txn` occurs here
   |                 borrow later used here
```

A short excerpt that reproduces the issue is [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=21755d50435029c0d1bb77702271cf23).

My guess is that the issue is created by returning `&'a mut Self` from `locked` etc, which isn't generally how things like that are implemented anyway - usually when I see patterns like that, they consume self and return Self. Changing from
```rust
pub fn locked(&'a mut self, locked: bool) -> &'a mut Self { ... }
let mut txn = LobbyTransaction::new();
txn.locked(true);
discord.update_lobby(id, txn, callback);
```
to
```rust
pub fn locked(mut self, locked: bool) -> Self { ... }
let txn = LobbyTransaction::new()
    .locked(true);
discord.update_lobby(id, txn, callback);
```
appears to fix the issue, as seen [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=49cc5947c488383591151a872f60070d). If you think this is a reasonable change, I'd be willing to change it myself and send in a pull request, for LobbyTransaction as well as whichever other types are structured the same way and would likely have the same problems.

**Comment by boringcactus** [comment]
On further reading, apparently the existing pattern is fairly widely used as well, including in the Rust standard library itself. So my proposed fix is probably not the best way to handle this, but I'm not quite sure what the right way to fix that would be.

**Comment by boringcactus** [comment]
Some help from strangers on the Internet suggests that the correct signature is
```rust
pub fn locked(&mut self, locked: bool) -> &mut Self
```
without the lifetime specifiers, which [more Rust Playground tinkering](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8b12827a076c89aea9a4349b4833493b) confirms. I think the issue is that `'a` denotes the metadata string lifetimes rather than the transaction itself, so making both the input and the output on `locked` obey the same lifetime causes problems. I'll probably start working on a pull request in the morning.

**Comment by ldesgoui** [comment]
Thanks for the report and the investigation, it seems I got my lifetimes confused, removing the lifetime annotation definitely fixes it, I will push a fix shortly.
I don't recall why I pass LobbyTransaction and similar by value, that probably can be changed to references, which would help with the builder pattern.
Feel free to ask for help using this crate, I'd be happy to help with anything

**Comment by ldesgoui** [comment]
The `fix-lifetimes` branch is available if you'd like to try it out
#37

**Comment by boringcactus** [comment]
that works! thanks a bunch

**Comment by ldesgoui** [comment]
great, published 0.3.1, thanks
